### PR TITLE
When the file does not exist, do not process tokens, when there is no tokens do not count null.

### PR DESCRIPTION
**Pull request by joejoseph00**
Use case: code is comming from the database, not a file.
for a bit of background please see views_php issue:
https://www.drupal.org/project/views_php/issues/2274543

I will sum it up:

The php code is comming from the database, not a file, closure library was prior to this commit trying to do a file_get_contents on this string: "path/to/drupal/and/module/path/views_php.module eval()'d code"
of course there is no filename called eval()'d code.

This ONLY occurs when using views_bulk_operations on views with a field containing views_php enabled code , I'm not sure why.  Normal use cases for views_php seem ok and do not try to load a file that doesn't exist.

### CODE REVIEW

**Comment by joejoseph00** [comment]
All passing

**Comment by msarca** [comment]
> All passing

That happened because you didn't provide any relevent unit tests and without them is really hard for me to figure out what exactly are you trying to solve. Please post a small snippet of code that will allow me to reproduce the issue without having to install the whole Drupal stack.

**Comment by joejoseph00** [comment]
In this use case of the closure library is only to replace the deprecated (in php 7.2.x) create_function

There was another workaround that was used but it had some issue serializing the closure if the code was cached.  So someone jammed in the closure library as follows:


the only functions provided by the closure library that is called by our module is:          


```
$function = views_php_create_function('$view, $plugin, $cache', $this->options['php_cache_results'] . ';');
ob_start();
$closure = $function->getClosure();
$closure($this->view, $this, $this->view->result);

/**
 * Internal support: create_function() emulation using an anonymous function with a serializable wrapper.
 */
function views_php_create_function($args, $code) {
  libraries_load('closure');
  return new \Opis\Closure\SerializableClosure(eval('return function(' . $args . ') {' . $code . '};'));
}
```


Where it gets complicated is this works fine without my patch in normal usage, however when using batch operations as provided by a third party module, the execution filename provided is referring to the file which contains this library call rather than it'self as is in normal use case where the views_php module executes the code directly.


all the closure code is provided not through files but loaded through the database.


So to reproduce this, I'd have to create a test that executes getClosure but call the test from another test using a batch operation.  Not sure how I'd go about doing this.


Bottom line is, the closure library should not be loading files that do not exist.  Should not process null tokens and should not count null tokens.  All of these three issues are resolved by this pull request without breaking any tests.

**Comment by joejoseph00** [comment]
So basically, the problem is somehow with another third party module that is maybe shovelling something invalid outside of the module that actually is using the closure library, rather than crash completely some null checks and file_exists prevents a catastrophic failure.  I don't really understand the other third party module nor the closure library.  The closure library was added to replace a php 7.2.x deprecated function called 'create_function'  , someone else had a different way but it didn't work with caching and serializing of certain types of code.  

I've been testing heavily with all the bits and pieces and have not yet noticed any regression with the pieces involved.  Sometimes you just have to hold your nose.

**Comment by joejoseph00** [comment]
however it seems to me that the closure library is doing something 'extra' that is not needed, all basically we need it for is to instantiate and call getClosure.  everything else might be 'bonus' and not needed at all.

**Comment by joejoseph00** [comment]
Is there any way to call getClosure for just arbitrary code that is loaded through a database?  so if I make one php file, but load getClosure 10 times with code provided by 10 different code rows  of a database , I wouldn't need to load the file that doesn't contain the code being executed because the code is from the database not from a file.

**Comment by joejoseph00** [comment]
basically we're doing this: instantiating SerializableClosure , and calling getClosure , why do we need a ReflectionClosure if we're just using SerializableClosure?

**Comment by sorinsarca** [comment]
>  I don't really understand the other third party module nor the closure library. The closure library was added to replace a php 7.2.x deprecated function called 'create_function' , someone else had a different way but it didn't work with caching and serializing of certain types of code.

It's true that you don't understand what this library does, but it also seems that you don't understand completely what `create_function` does. It is possible to emulate `create_function`
with opis/closure, but not the way you are trying to. We also needed something more advanced that you need, and luckily we just have it.

Here is an replacement example of your `views_php_create_function` using the [opis/stream](https://github.com/opis/stream) library

```php
function views_php_create_function($args, $code) {
   $isPHPAtLeast72 = true; // just check the current php version
    
  if ($isPHPAtLeast72) {
      // perhaps you must load the opis/stream lib somehow: libraries_load('stream');
      return \Opis\Stream\Wrapper\PHPCodeStreamWrapper::include('<?php return function(' . $args . ') {' . $code. '};');
  }

  // otherwise use the "non-deprecated" create_function

  return create_function($args, $code);
}
```

**Comment by mustanggb** [comment]
Just to chime in here, thanks a lot sorinsarca, this looks like a much better option than a workaround to a workaround to a workaround that was being attempted before.

No demands, obviously, but it seems like quite a few people would be interested in opis/stream, however the docs seems quite sparse.

**Comment by sorinsarca** [comment]
> No demands, obviously, but it seems like quite a few people would be interested in opis/stream

Great, then don't use it. 

Off-topic: I think that composer should add a `--stars` flag, to prevent it from downloading projects that don't have a minimum number of github stars.

[On-topic](https://www.php.net/manual/en/language.oop5.basic.php#language.oop5.basic.extends)

**Comment by joejoseph00** [comment]
I will try the opis/stream library asap and report back the findings. Meanwhile the closure library works in all our known use cases provided this pull request is used.

**Comment by msarca** [comment]
Hey, @joejoseph00 & @mustanggb! I've finally managed to take a closer look at your issue and I came up with a general solution that can be used by all those who are looking to emulate `create_function`'s behavior.  Sorry for taking so long, but we are in crunch mode for over a month now, and the effects are starting to show.

All you will have to do is to add a single extra line of code, like below

```php
/**
 * Internal support: create_function() emulation using an anonymous function with a serializable wrapper.
 */
function views_php_create_function($args, $code) {
  libraries_load('closure');
  $closure = \Opis\Closure\SerializableClosure::createClosure($args, $code);
  return new \Opis\Closure\SerializableClosure($closure);
}
```
The static `createClosure` method was added in version 3.4.0(just released), so issuing a `composer update` command is required first. Try applying the suggested changes and give me some feedback.

**Comment by joejoseph00** [comment]
hi @msarca  I took your advice, grabbed the latest closure library, adjusted and refactored all code, however now I get this:

```
Parameter 3 to Opis\Closure\SerializableClosure::{closure}() expected to be a reference, value given in Opis\Closure\SerializableClosure->__invoke() (line 109 of /path/to/ibraries/closure/src/SerializableClosure.php
```

here's my patch:


```
 function views_php_create_function($args, $code) {
   libraries_load('closure');
  $closure = \Opis\Closure\SerializableClosure::createClosure($args, $code);
  return new \Opis\Closure\SerializableClosure($closure);
 }
```


have a look here, I call the views_php_create_function  with two parameters.

```
      $function = views_php_create_function('$view, $handler, &$static', $this->options['php_setup'] . ';');
      ob_start();
      $function($this->view, $this, $this->php_static_variable);
      ob_end_clean();
```

however with the instantiated static instance of SerializableClosure , how do I pass something in , if it has one or many parameters, in this case there's three parameters?

Something I'm not getting here, do I still call getClosure()???



it worked before when I was using getClosure ..


Anyhow, it was working now its not working.
here is NEW approach
[views_php-new_approach-2274543-92.patch.txt](https://github.com/opis/closure/files/3567282/views_php-new_approach-2274543-92.patch.txt)

Here is the original approach which led to this pull request:
[views_php-2274543-81_ORIGINAL_APPROACH.patch.txt](https://github.com/opis/closure/files/3567283/views_php-2274543-81_ORIGINAL_APPROACH.patch.txt)

**Comment by joejoseph00** [comment]
so ya this new approach isn't exactly a "drop-in" replacement for create_function as I attempted in the new approach patch (see above).
maybe I am missing something.

**Comment by msarca** [comment]
> Parameter 3 to Opis\Closure\SerializableClosure::{closure}() expected to be a reference, value given in Opis\Closure\SerializableClosure->__invoke() (line 109 of /path/to/ibraries/closure/src/SerializableClosure.php

This happens because you are invoking the `SerializableClosure` wrapper directly and your constructed closure has a 3rd parameter which gets is value by reference. By invoking the wrapper directly you cannot pass values by reference and hence this error ocurres. 

> maybe I am missing something.

It's a matter of choice.. You must decide over which data type will be returned by `views_php_create_function`. Do you want to return a wrapper or a closure? If you return a wrapper, then you must obtain the wrapped closure by calling `getClosure`. Otherwise, if you want to return a closure, the below code will suffice.

```php
 function views_php_create_function($args, $code) {
   libraries_load('closure');
  return \Opis\Closure\SerializableClosure::createClosure($args, $code);
 }
```

**Comment by msarca** [comment]
As like the previous PR, this one also treats the effects, not the cause. And the cause is the usage of `eval` function for creating closures on the fly. With the introduction of `createClosure` method, the cause can be eliminated now.

```php
$closure = SerializableClosure::createClosure('$a, $b', 'return $a + $b;');
echo $closure(3, 4), PHP_EOL; //> 7

// The closure can now be wrapped and serialized
$wrapper = serialize(new SerializableClosure($closure));

// Then unserialized
$closure = unserialize($wrapper)->getClosure();

// .. and used again
echo $closure(7, 4), PHP_EOL; //> 11
```

**Comment by joejoseph00** [comment]
@MustangGB can you throw me a bone here?

**Comment by joejoseph00** [comment]
actually, I did manage to get it to autoload, but scrapping this as an option because it results in this regression:
Error message

`Exception: Serialization of 'Closure' is not allowed in serialize() (line 465 of /web/includes/cache.inc)`

**Comment by joejoseph00** [comment]
So, the only option that works for us is the closure library with the patch on file_exists , and null checks for tokens and preventing count of null tokens.

Thanks for your good intentions and suggestions however I will be recommending my forked version of this project going forward.

here is my forked repo:
https://github.com/entreprise7pro/closure

**Comment by sorinsarca** [comment]
> `Exception: Serialization of 'Closure' is not allowed in serialize() (line 465 of /web/includes/cache.inc)`

This is because you don't use the `SerializableClosure` wrapper, given that you want to serialize closures.

```php
function views_php_create_function($args, $code) {
   libraries_load('closure');
  
  $plain = \Opis\Closure\SerializableClosure::createClosure($args, $code);
  $serializable = \Opis\Closure\SerializableClosure::from($plain);

  return $serializable;
 }
```

**Comment by joejoseph00** [comment]
Thanks again for the suggestion, I did try exactly that and it gives me the same error about passing a reference.

```Warning: Parameter 3 to Opis\Closure\SerializableClosure::{closure}() expected to be a reference, value given in Opis\Closure\SerializableClosure->__invoke() (line 109 of web/sites/all/libraries/closure/src/SerializableClosure.php).```

**Comment by joejoseph00** [comment]
the only thing that works for me is using eval as follows:

```
/**
 * Internal support: create_function() emulation using an anonymous function with a serializable wrapper.
 */
function views_php_create_function($args, $code) {
  libraries_load('closure');
  return new \Opis\Closure\SerializableClosure(eval('return function(' . $args . ') {' . $code . '};'));
}

// Called something like this:
      $function = views_php_create_function('$view, $handler, &$static', $this->options['php_setup'] . ';');
      ob_start();
      $closure = $function->getClosure();
      $closure($this->view, $this, $this->php_static_variable);
      ob_end_clean();

```

**Comment by joejoseph00** [comment]
the above example works with my forked version of the opis/closure library https://github.com/entreprise7pro/closure

**Comment by joejoseph00** [comment]
so there's a couple of approaches, the non eval one that you suggested I made it a drop-in replacement for create_function

however the one that works is not a drop-in replacement for create_function it's returning a closure then I'm adding the parameters to that closure in the caller, so two steps.  This approach due to the third party module invoking a batch operation requires that I patch the closure library for file_exists and do some validation on null tokens.

**Comment by joejoseph00** [comment]
by it'self the closure library doesn't need to be patched, it's only when I start using batch operations invoked by a third party module.

**Comment by msarca** [comment]
> Thanks again for the suggestion, I did try exactly that and it gives me the same error I reported the other day about expecting a reference.

And I've told you that happens because you are invoking the wrapper directly.. So you need to get the wrapped closure, and invoke that instead.

```php
$closure = $function->getClosure();
$closure($this->view, $this, $this->php_static_variable);
```

**Comment by joejoseph00** [comment]
@msarca ok yes that works if I use the getClosure()

**Comment by joejoseph00** [comment]
thanks yes, I confirm this works without the file_exists checks, thanks for your insistence and guidance on this!

**Comment by joejoseph00** [comment]
@msarca  thanks again, here's the interdiff that seems to work well using your suggestion.

https://www.drupal.org/files/issues/2019-09-06/interdiff-2274543-81_to_95.txt

**Comment by mustanggb** [comment]
@msarca Bit late to the party here, but it works lovely, thanks a lot for your expert guidance, and the wrapper function.
